Causes compilation to fail with contravariant typeclasses

The following example causes compilation to fail:

```scala
@typeclass trait RowWriter[-A] { self =>
  def write(a: A): Seq[String]
}
```

The compilation error is:
```
[trace] Stack trace suppressed: run last compile:compile for the full output.
[error] (compile:compile) java.lang.AssertionError: assertion failed: Trying to create a method with variant type parameters: (List(type A),(implicit instance: com.nrinaudo.csv.RowWriter[A])com.nrinaudo.csv.RowWriter[A])
```

I have reproduced this using:

* simulacrum `0.4.0`
* macroparadise `2.1.0-M5`
* scala `2.11.7`

--- a/core/src/main/scala/simulacrum/typeclass.scala
+++ b/core/src/main/scala/simulacrum/typeclass.scala
@@ -55,6 +55,16 @@ class TypeClassMacros(val c: Context) {
 
   def generateTypeClass(annottees: c.Expr[Any]*): c.Expr[Any] = {
 
+    /** Can't have variant modifiers on method type parameters.
+     */
+    def eliminateVariance(tparam: TypeDef): TypeDef = {
+      // If there's another way to do this I'm afraid I don't know it.
+      val u        = c.universe.asInstanceOf[c.universe.type with scala.reflect.internal.SymbolTable]
+      val tparam0  = tparam.asInstanceOf[u.TypeDef]
+      val badFlags = (Flag.COVARIANT | Flag.CONTRAVARIANT).asInstanceOf[Long]
+      u.copyTypeDef(tparam0)(mods = tparam0.mods & ~badFlags)
+    }
+
     def trace(s: => String) = {
       // Macro paradise seems to always output info statements, even without -verbose
       if (sys.props.get("simulacrum.trace").isDefined) c.info(c.enclosingPosition, s, false)
@@ -235,7 +245,7 @@ class TypeClassMacros(val c: Context) {
 
     def generateOps(typeClass: ClassDef, tcInstanceName: TermName, tparam: TypeDef, proper: Boolean, liftedTypeArg: Option[TypeDef]): (ClassDef, Set[TypeName]) = {
       val adaptedMethods = adaptMethods(typeClass, tcInstanceName, tparam.name, proper, liftedTypeArg)
-      val tparams = List(tparam) ++ liftedTypeArg
+      val tparams = List(eliminateVariance(tparam)) ++ liftedTypeArg
       val tparamNames = tparams.map { _.name }
       val targetType = liftedTypeArg.map(lta => tq"${tparam.name}[${lta.name}]").getOrElse(tq"${tparam.name}")
 
@@ -272,7 +282,8 @@ class TypeClassMacros(val c: Context) {
       }"""
     }
 
-    def generateCompanion(typeClass: ClassDef, tparam: TypeDef, proper: Boolean, comp: Tree) = {
+    def generateCompanion(typeClass: ClassDef, tparam0: TypeDef, proper: Boolean, comp: Tree) = {
+      val tparam = eliminateVariance(tparam0)
       val summoner = q"def apply[$tparam](implicit instance: ${typeClass.name}[${tparam.name}]): ${typeClass.name}[${tparam.name}] = instance"
 
       val liftedTypeArg = if (proper) None else Some {
@@ -304,7 +315,7 @@ class TypeClassMacros(val c: Context) {
       val allOpsTrait = generateAllOps(typeClass, tcInstanceName, tparam, liftedTypeArg)
 
       def generateOpsImplicitConversion(opsType: TypeName, methodName: TermName) = {
-        val tparams = List(tparam) ++ liftedTypeArg
+        val tparams = List(eliminateVariance(tparam)) ++ liftedTypeArg
         val tparamNames = tparams.map(_.name)
         val targetType = liftedTypeArg.map(lta => tq"${tparam.name}[${lta.name}]").getOrElse(tq"${tparam.name}")
         q"""

--- a/core/src/test/scala/simulacrum/typeclass.scala
+++ b/core/src/test/scala/simulacrum/typeclass.scala
@@ -290,5 +290,34 @@ class TypeClassTest extends WordSpec with Matchers {
       @typeclass trait Foo[F[_]]
       @exports object Foo
     }
+
+    /** It'd be a lot better to parameterize all the tests over the variance
+     *  of the type parameter so they can all be run again, but I"m not sure
+     *  how to do it.
+     */
+    "support variant type classes" which {
+      "are covariant" in {
+        @typeclass trait Empty[+A] {
+          def empty: A
+        }
+        @typeclass trait Empty1[+A] extends Empty[A] {
+          def empty2[A1 >: A]: A1
+        }
+        @typeclass trait Empty2[A] extends Empty[A] {  // invariant
+          def id(x: A): A
+        }
+      }
+      "are contravariant" in {
+        @typeclass trait Order[-A] {
+          def cmp(x: A, y: A): Boolean
+        }
+        @typeclass trait Order1[-A] extends Order[A] {
+          def cmp2[A1 <: A](x: A1, y: A1): Boolean
+        }
+        @typeclass trait Order2[A] extends Order[A] { // invariant
+          def id(x: A): A
+        }
+      }
+    }
   }
 }

